Add EKS Guide (#278)
scalar-labs/scalar-terraform: docs/EKSGuide.md

> # Sidenotes on How to create a Kubernetes cluster in Amazon EKS with scalar-terraform

Reviewer: The capitalization of the headings should be consistent.

```suggestion
# Sidenotes on how to create a Kubernetes cluster in Amazon EKS with scalar-terraform
```

## Manage bastion role for an EKS cluster

To grant bastion role the ability to interact with your cluster, you must edit the aws-auth ConfigMap within Kubernetes.

> By default, `scalar-terraform` will automatically create when `public_cluster_enabled` is set true (true by default).

Reviewer 2: What does it create? (It doesn't have the object.)

Author: The `aws-auth ConfigMap` will be created.
For details,

- To apply the aws-auth ConfigMap to your cluster
- To add an IAM bastion role to your cluster

How about adding `the aws-auth ConfigMap`?

e.g.,
`・・・will automatically create the aws-auth ConfigMap when ・・・`

## Manage Cluster Autoscaler for node autoscaling

The Cluster Autoscaler minimizes costs by ensuring that nodes are only added to the cluster when needed and are removed when unused.
With `cluster_auto_scaling = "true"`,  `scalar-terraform` automatically deploys Autoscaler when `public_cluster_enabled` is set true (true by default).

```terraform
region = "ap-northeast-1"

kubernetes_cluster = {
  cluster_auto_scaling = "true"
}
```

If you want to add Autoscaler manually, see [Deploy the Cluster Autoscaler](https://docs.aws.amazon.com/eks/latest/userguide/cluster-autoscaler.html#ca-deploy).

## Update EKS cluster endpoint access to `Private`

To disable public access for your cluster's Kubernetes API server endpoint, please follow the steps below.

- Edit [example.tfvars](https://github.com/scalar-labs/scalar-terraform-examples/blob/main/aws/kubernetes/example.tfvars)

```terraform
region = "ap-northeast-1"

kubernetes_cluster = {
  public_cluster_enabled = "false"
}
```

- Terraform apply

```console
cd aws/kubernetes
terraform init
terraform apply -var-file example.tfvars

> ```

Reviewer: These code blocks in the list items should be indented with 4 spaces so that the list is rendered to one `<ul>` tag.

Author: Added 4 spaces.

# Amazon EKS Guide

## [Amazon EKS cluster endpoint access control](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html)

By default, `scalar-terraform` will create eks cluster with `Public and private` endpoint access.
It is recommended to enable both public and private endpoint access by `public_cluster_enabled = "true"` when first build an environment.
This is an important prerequisite for creating [aws-auth ConfigMap](#managing-bastion-role-for-eks-cluster) and [Autoscaler](#managing-cluster-autoscaler-for-node-autoscaling) with `scalar-terraform`.

```terraform
region = "ap-northeast-1"

kubernetes_cluster = {
  cluster_endpoint_public_access_cidrs = "0.0.0.0/0"
  public_cluster_enabled               = "true"
}
```

## Managing bastion role for EKS cluster

To grant bastion role the ability to interact with your cluster, you must edit the aws-auth ConfigMap within Kubernetes.

> By default, `scalar-terraform` will automatically create when you use the public access endpoint.

Reviewer 2: ```suggestion
By default, `scalar-terraform` will automatically create when `public_cluster_enabled` is set true (true by default).
```
Correct?

If you want to add permissions manually, see [Managing users or IAM roles for your cluster](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html).

## Managing Cluster Autoscaler for node autoscaling

The Cluster Autoscaler minimizes costs by ensuring that nodes are only added to the cluster when needed and are removed when unused.

> To enable autoscaling with `cluster_auto_scaling = "true"` and `scalar-terraform` will automatically deploy autoscaler when you use the public access endpoint.

Reviewer 2: ```suggestion
With `cluster_auto_scaling = "true"`,  `scalar-terraform` automatically deploys Autoscaler when `public_cluster_enabled` is set true (true by default).
```

```terraform
region = "ap-northeast-1"

kubernetes_cluster = {
  cluster_auto_scaling = "true"
}
```

> If you want to add autoscaler manually, see [Deploy the Cluster Autoscaler](https://docs.aws.amazon.com/eks/latest/userguide/cluster-autoscaler.html#ca-deploy).

Reviewer 2: ```suggestion
If you want to add Autoscaler manually, see [Deploy the Cluster Autoscaler](https://docs.aws.amazon.com/eks/latest/userguide/cluster-autoscaler.html#ca-deploy).
```

## Update EKS cluster endpoint access to `Private`

> To enable private access to the Kubernetes API server so that all communication between your nodes and the API server stays within your VPC.

Reviewer 2: ```suggestion
To disable public access for your cluster's Kubernetes API server endpoint, please follow the steps below.
```

> Any `kubectl` commands must come from within the VPC or a connected network. For connectivity options, see [Accessing a private only API server](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html#private-access).

Reviewer 2: ```suggestion
```
Is this needed?

- Edit [example.tfvars](https://github.com/scalar-labs/scalar-terraform-examples/blob/main/aws/kubernetes/example.tfvars)

```terraform
region = "ap-northeast-1"

kubernetes_cluster = {
  public_cluster_enabled = "false"
}
```

- Terraform apply

```console
cd aws/kubernetes
terraform init
terraform apply -var-file example.tfvars
```

## Update EKS cluster endpoint access from `Private` to `Public and private`

> Because `scalar-terraform` does not support yet, so please change from the AWS console, AWS CLI, see [Modifying cluster endpoint access](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html#modify-endpoint-access).

Reviewer 2: ```suggestion
Since `scalar-terraform` does not support the flow yet, please change it from the AWS console, AWS CLI, see [Modifying cluster endpoint access](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html#modify-endpoint-access).
```
